runtime/eventring.c

```diff
@@ -0,0 +1,549 @@
+    to run [teardown_eventring] first. */
+    int remove_file = 0;
+    do {
+      caml_try_run_on_all_domains(&stw_teardown_eventring, &remove_file, NULL);
+    } while(atomic_load_acq(&eventring_enabled));
+
+    /* We still have the path and ring size from our parent */
+    caml_eventring_start();
+  }
+}
+
+/* Return the current location for the ring buffers of this process. This is
+  used in the consumer to read the ring buffers of the current process */
+char_os* caml_eventring_current_location() {
+  if( atomic_load_acq(&eventring_enabled) ) {
+    return current_ring_loc;
+  } else {
+    return NULL;
+  }
+}
+
+/* Write a lifecycle event and then trigger a stop the world to tear down the
+  ring buffers */
+void caml_eventring_destroy() {
+  if (atomic_load_acq(&eventring_enabled)) {
+    write_to_ring(EV_RUNTIME, EV_LIFECYCLE, EV_RING_STOP, 0, NULL, 0);
+
+    /* clean up eventring when we exit. */
+    int remove_file = 1;
+    do {
+      caml_try_run_on_all_domains(&stw_teardown_eventring, &remove_file, NULL);
+    }
+    while( atomic_load_acq(&eventring_enabled) );
+  }
+}
+
+/* Create the ring buffers. This must only be called from a stop the world
+  pause. */
+static void
+stw_create_eventring(caml_domain_state *domain_state, void *data,
+                              int num_participating,
+                              caml_domain_state **participating_domains) {
+  /* Everyone must be stopped for starting and stopping eventring */
+  caml_global_barrier();
+
+  /* Only do this on one domain */
+  if (participating_domains[0] == domain_state) {
+    /* Don't initialise eventring twice */
+    if (!atomic_load_acq(&eventring_enabled)) {
+      int ret, ring_headers_length;
+#ifdef _WIN32
+      DWORD pid = GetCurrentProcessId();
+#else
+      int ring_fd;
+      long int pid = getpid();
+#endif
+
+      current_ring_loc = caml_stat_alloc(EVENTRING_MAX_MSG_LENGTH);
+
+      if (eventring_path) {
+        snprintf_os(current_ring_loc, EVENTRING_MAX_MSG_LENGTH,
+                    T("%s/%ld.eventring"), eventring_path, pid);
+      } else {
+        snprintf_os(current_ring_loc, EVENTRING_MAX_MSG_LENGTH,
+                    T("%ld.eventring"), pid);
+      }
+
+      current_ring_total_size =
+          Max_domains * (ring_size_words * sizeof(uint64_t) +
+                         sizeof(struct eventring_buffer_header)) +
+          sizeof(struct eventring_metadata_header);
+
+#ifdef _WIN32
+      ring_file_handle = CreateFile(
+        current_ring_loc,
+        GENERIC_READ | GENERIC_WRITE,
+        FILE_SHARE_READ | FILE_SHARE_WRITE,
+        NULL,
+        CREATE_ALWAYS,
+        FILE_ATTRIBUTE_NORMAL,
+        NULL
+      );
+
+      if (ring_file_handle == INVALID_HANDLE_VALUE) {
+        char* ring_loc_u8 = caml_stat_strdup_of_os(current_ring_loc);
+        caml_fatal_error("Couldn't open ring buffer loc: %s",
+                          ring_loc_u8);
+        caml_stat_free(ring_loc_u8);
```

Comment:
It's weird to call `caml_stat_free` after `caml_fatal_error`.

Reply:
It is. I mostly left that as documentation in case someone decides to make that non-fatal.